File: utils/ManualOperation/catkin_ws/src/motors_driver/src/motor_controller.cpp
#include "ros/ros.h"
#include "std_msgs/Int32MultiArray.h"
#include <sstream>
#include <geometry_msgs/Twist.h>

class MotorController
{
    public:
        MotorController();
    private:
        void controllerCallback(const sensor_msgs::Joy::ConstPtr& joy);

        ros::NodeHandle node_;
        ros::Publisher motor_pub_;
        ros::Subscriber twist_sub_;

        std_msgs::Int32MultiArray msg;
}

MotorController::MotorController()
{
    ros::NodeHandle node_;
    motor_pub_ = node_.advertise<std_msgs::Int32MultiArray>("command", 100);
    twist_sub_ = node_.subscribe<geometry_msgs::Twist>("motors/motor_twist", 100, &MotorController::controllerCallback, this);

    msg = std_msgs::Int32MultiArray();
    msg.layout.dim.push_back(std_msgs::MultiArrayDimension());
    msg.layout.dim[0].label = "command";
    msg.layout.dim[0].size = 16;
    msg.layout.dim[0].stride = 16;
}

void PS4ControllerMotor::controllerCallback(const geometry_msgs::Twist::ConstPtr& twist)
{
    // geometry_msgs::Twist twist;
    // twist.linear.x = 10*joy->axes[1];
    // twist.angular.z = 10*joy->axes[2];
    // motor_pub_.publish(twist);

    // ROS_INFO("X: [%f], Y: [%f]", joy->axes[1], joy->axes[2]);
    // ROS_INFO("X: [%f], Y: [%f]", twist.linear.x, twist.angular.z);
    msg.data = std::vector<int>{count, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1};
    // ROS_INFO("[Talker] I published %d\n", count);
    // chatter_publisher.publish(msg);
}

int main(int argc, char **argv){
	ros::init(argc, argv, "motor_controller");
    
    MotorController motor_controller;
    ros::spin();
    
    return 0;
}

File: utils/ManualOperation/catkin_ws/src/motors_driver/scripts/motor_controller.py
#!/usr/bin/python

import rospy
from std_msgs.msg import Int32MultiArray
from std_msgs.msg import MultiArrayDimension
from geometry_msgs.msg import Twist

class MotorController:
    def __init__(self):
        self.pub = rospy.Publisher('command', Int32MultiArray, queue_size=1000)
        self.sub = rospy.Subscriber("motors/motor_twist", Twist, self.callback)
        self.msg = Int32MultiArray()
        self.msg.layout.dim.append(MultiArrayDimension())
        self.msg.layout.dim[0].label = "command"
        self.msg.layout.dim[0].size = 16
        self.msg.layout.dim[0].stride = 16

        self.vel_zero = 4900
        self.dir_zero = 4500

        self.vel_max = 300
        self.dir_max = 2000

        # self.msg.data = [4900, 4900, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1]

    
    def callback(self, twist):
        if twist.linear.x >= 0:
            x_val = int(self.vel_zero + (twist.linear.x * self.vel_max))
        else:
            x_val = int(self.vel_zero + (twist.linear.x * self.vel_max) - 200)
        z_val = int(self.dir_zero + (twist.angular.z * self.dir_max))
        self.msg.data = [x_val, z_val, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1]

        # print(twist.linear.x)
        print(twist.angular.z)
        print(self.msg.data)
        self.pub.publish(self.msg)


def main():
    rospy.init_node('motor_controller')
    obc = MotorController()
    try:
        rospy.spin()
    except KeyboardInterrupt:
        print("Shutting down")


if __name__ == '__main__':
    main()

# def talker():
    
#     pub = rospy.Publisher('command', Int32MultiArray, queue_size=1000)
#     rate = rospy.Rate(0.5) # 10hz
    
#     msg = Int32MultiArray()
#     print(msg)
#     msg.layout.dim.append(MultiArrayDimension())
#     msg.layout.dim[0].label = "command"
#     msg.layout.dim[0].size = 16
#     msg.layout.dim[0].stride = 16
    
#     val90 = [4900, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1]
#     val20 = [2900, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1]
#     val180 = [6900, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1]
    
#     velocity0 = [-1, 4900, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1]
#     velocityF = [-1, 2900, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1]
#     velocityB = [-1, 6900, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1]
    
# #     msg.data = [6900, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1]
    
#     counter = 0
#     while not rospy.is_shutdown():
# #         hello_str = "hello world %s" % rospy.get_time()
        
#         if counter == 0:
#             msg.data = val90
#         elif counter == 1:
#             msg.data = val20
#         elif counter == 2:
#             msg.data = val180
            
# #         if counter == 0:
# #             msg.data = velocity0
# #         elif counter == 1:
# #             msg.data = velocityF
# #         elif counter == 2:
# #             msg.data = velocity0
            
#         rospy.loginfo("sending")
#         print("{}".format(counter))
#         pub.publish(msg)
#         rate.sleep()
#         counter += 1
#         counter = counter % 3
